nav2_costmap_2d/src/costmap_2d_ros.cpp

```diff
@@ -35,338 +35,406 @@
-  int map_width_meters, map_height_meters;
+nav2_lifecycle::CallbackReturn
+Costmap2DROS::on_cleanup(const rclcpp_lifecycle::State & /*state*/)
+{
+  RCLCPP_INFO(get_logger(), "Cleaning up");
 
-  dynamic_param_client_->get_event_param("width", map_width_meters);
-  dynamic_param_client_->get_event_param("height", map_height_meters);
-  dynamic_param_client_->get_event_param("resolution", resolution);
-  dynamic_param_client_->get_event_param("origin_x", origin_x);
-  dynamic_param_client_->get_event_param("origin_y", origin_y);
+  tf_buffer_.reset();
+  tf_listener_.reset();
 
-  if (!layered_costmap_->isSizeLocked()) {
-    layered_costmap_->resizeMap((unsigned int)(map_width_meters / resolution),
-      (unsigned int)(map_height_meters / resolution), resolution, origin_x, origin_y);
+  footprint_sub_.reset();
+  footprint_pub_.reset();
+
+  if (costmap_publisher_ != nullptr) {
+    delete costmap_publisher_;
+    costmap_publisher_ = nullptr;
   }
 
-  // If the padding has changed, call setUnpaddedRobotFootprint() to
-  // re-apply the padding.
-  float footprint_padding;
-  dynamic_param_client_->get_event_param("footprint_padding", footprint_padding);
+  delete layered_costmap_;
+  layered_costmap_ = nullptr;
 
-  if (footprint_padding_ != footprint_padding) {
-    footprint_padding_ = footprint_padding;
-    setUnpaddedRobotFootprint(unpadded_footprint_);
-  }
+  clear_costmap_service_.reset();
 
-  readFootprintFromConfig();
+  costmap_update_monitor_.reset();
 
-  map_update_thread_ = new std::thread(std::bind(
-        &Costmap2DROS::mapUpdateLoop, this, map_update_frequency));
+  return nav2_lifecycle::CallbackReturn::SUCCESS;
 }
 
-void Costmap2DROS::readFootprintFromConfig()
+nav2_lifecycle::CallbackReturn
+Costmap2DROS::on_error(const rclcpp_lifecycle::State &)
 {
-  // Only change the footprint if footprint or robot_radius has
-  // changed.  Otherwise we might overwrite a footprint sent on a
-  // topic by a dynamic_reconfigure call which was setting some other
-  // variable.
+  RCLCPP_ERROR(get_logger(), "Handling error state");
+  return nav2_lifecycle::CallbackReturn::SUCCESS;
+}
 
-  std::string footprint;
-  double robot_radius;
-  dynamic_param_client_->get_event_param("footprint", footprint);
-  dynamic_param_client_->get_event_param("robot_radius", robot_radius);
+nav2_lifecycle::CallbackReturn
+Costmap2DROS::on_shutdown(const rclcpp_lifecycle::State &)
+{
+  RCLCPP_INFO(get_logger(), "Shutting down");
+  return nav2_lifecycle::CallbackReturn::SUCCESS;
+}
 
-  if (footprint != "" && footprint != "[]") {
+void
+Costmap2DROS::getParameters()
+{
+  RCLCPP_DEBUG(get_logger(), " getParameters");
+
+  std::vector<std::string> plugin_names{"static_layer", "obstacle_layer", "inflation_layer"};
+  std::vector<std::string> plugin_types{"nav2_costmap_2d::StaticLayer", "nav2_costmap_2d::ObstacleLayer", "nav2_costmap_2d::InflationLayer"};
+
+  declare_parameter("always_send_full_costmap", rclcpp::ParameterValue(false));
```

Comment:
I don't think we should be declaring parameters here if this function is going to be used in the `on_configure` step as this would cause an error I believe. We should have parameters declared only once (in constructor I think).

Reply:
Yes, you're absolutely right. Moved to the constructor.